src/System.Diagnostics.Process/src/System/Diagnostics/AsyncStreamReader.cs

```diff
@@ -148,76 +87,67 @@ internal void BeginReadLine()
-
-            try
-            {
-                byteLen = t.GetAwaiter().GetResult();
-            }
-            catch (IOException)
-            {
-                // We should ideally consume errors from operations getting cancelled
-                // so that we don't crash the unsuspecting parent with an unhandled exc. 
-                // This seems to come in 2 forms of exceptions (depending on platform and scenario), 
-                // namely OperationCanceledException and IOException (for errorcode that we don't 
-                // map explicitly).   
-                byteLen = 0; // Treat this as EOF
-            }
-            catch (OperationCanceledException)
+            while (true)
             {
-                // We should consume any OperationCanceledException from child read here  
-                // so that we don't crash the parent with an unhandled exc
-                byteLen = 0; // Treat this as EOF
-            }
+                int byteLen;
 
-            if (byteLen == 0)
-            {
-                // We're at EOF, we won't call this function again from here on.
-                lock (_messageQueue)
+                try
                 {
-                    if (_sb.Length != 0)
-                    {
-                        _messageQueue.Enqueue(_sb.ToString());
-                        _sb.Length = 0;
-                    }
-                    _messageQueue.Enqueue(null);
+                    byteLen = await _stream.ReadAsync(_byteBuffer, 0, _byteBuffer.Length, _cts.Token).ConfigureAwait(false);
                 }
-
-                try
+                catch (IOException)
                 {
-                    // UserCallback could throw, we should still set the eofEvent 
-                    FlushMessageQueue();
+                    // We should ideally consume errors from operations getting cancelled
+                    // so that we don't crash the unsuspecting parent with an unhandled exc.
+                    // This seems to come in 2 forms of exceptions (depending on platform and scenario),
+                    // namely OperationCanceledException and IOException (for errorcode that we don't
+                    // map explicitly).
+                    byteLen = 0; // Treat this as EOF
                 }
-                finally
+                catch (OperationCanceledException)
                 {
-                    _eofEvent.Set();
+                    // We should consume any OperationCanceledException from child read here
+                    // so that we don't crash the parent with an unhandled exc
+                    byteLen = 0; // Treat this as EOF
                 }
-            }
-            else
-            {
+
+                if (byteLen == 0)
+                {
+                    // We're at EOF, we won't start the next iteration so we process current buffer content,
+                    // flush message queue and return
+                    lock (_messageQueue)
```

Comment:
>it'd be cleaner to just move all of this to after the while loop. then the catch blocks would just "break;", in the try above we'd also break if byteLen == 0

agree

>and the logic around getting chars / appending / moving lines would also be moved into the try.

disagree, because `MoveLinesFromStringBuilderToMessageQueue()` calls up `FlushMessageQueue()` and that in turn calls user's delegate which could throw.